File: cli/agentscale/commands/validate.py
"""Validate command - Validate agent before deployment."""

from pathlib import Path
import typer
import yaml

from agentscale.utils.output import print_error, print_info, print_success, print_warning


def validate(
    agent_path: str = typer.Argument(..., help="Path to agent directory"),
) -> None:
    """Validate agent configuration before deployment.

    Checks agent.yaml structure, module files, and requirements.txt
    to catch errors before deploying.

    Examples:
        agentscale validate ./my-agent
        agentscale validate ../calculator-agent
    """
    agent_dir = Path(agent_path)

    print_info(f"Validating agent at {agent_path}...")
    print("")

    errors = []
    warnings = []

    # Check 1: Directory exists
    if not agent_dir.exists():
        print_error("Agent directory not found", f"Path: {agent_path}")
        raise typer.Exit(1)

    if not agent_dir.is_dir():
        print_error("Agent path must be a directory", f"Path: {agent_path}")
        raise typer.Exit(1)

    # Check 2: agent.yaml exists
    agent_yaml = agent_dir / "agent.yaml"
    if not agent_yaml.exists():
        errors.append("agent.yaml not found")
    else:
        print_success("✓ agent.yaml found")

        # Check 3: Valid YAML
        try:
            config = yaml.safe_load(agent_yaml.read_text())

            # Check 4: Required fields
            required_fields = ["name", "runtime", "module", "entrypoint"]
            missing = [f for f in required_fields if f not in config]

            if missing:
                errors.append(f"Missing required fields: {', '.join(missing)}")
            else:
                print_success(f"✓ Required fields present")

            # Check 5: Module file exists
            if "module" in config:
                module_name = config["module"].rstrip(".py")
                module_file = agent_dir / f"{module_name}.py"

                if module_file.exists():
                    print_success(f"✓ Module file exists: {module_name}.py")
                else:
                    errors.append(f"Module file not found: {module_name}.py")

            # Check 6: Entrypoint specified
            if "entrypoint" in config:
                print_success(f"✓ Entrypoint function: {config['entrypoint']}")

            # Check 7: Runtime supported
            supported_runtimes = ["python3", "nodejs20"]
            if "runtime" in config and config["runtime"] not in supported_runtimes:
                errors.append(f"Unsupported runtime: {config['runtime']} (supported: {', '.join(supported_runtimes)})")

        except yaml.YAMLError as e:
            errors.append(f"Invalid YAML: {e}")

    # Check 8: requirements.txt (optional)
    req_file = agent_dir / "requirements.txt"
    if req_file.exists():
        try:
            reqs = req_file.read_text().strip().split("\n")
            reqs = [r.strip() for r in reqs if r.strip() and not r.startswith("#")]

            if reqs:
                print_success(f"✓ requirements.txt valid ({len(reqs)} package(s))")

                # Warn about large dependencies
                large_deps = ["tensorflow", "torch", "transformers", "opencv"]
                for dep in reqs:
                    dep_name = dep.split("==")[0].split(">=")[0].split("<=")[0].lower()
                    if any(large in dep_name for large in large_deps):
                        warnings.append(f"Large dependency detected: {dep_name} (may increase image size)")

            else:
                warnings.append("requirements.txt is empty")

        except Exception as e:
            warnings.append(f"Error reading requirements.txt: {e}")

    # Display summary
    print("")

    if errors:
        print_error("Validation failed", f"{len(errors)} error(s) found:")
        for error in errors:
            print(f"  ✗ {error}")
        print("")
        raise typer.Exit(1)

    if warnings:
        for warning in warnings:
            print_warning("⚠", warning)
        print("")

    print_success("Agent is ready to deploy!")
    print("")
    print("Deploy with:")
    print(f"  agentscale deploy {agent_path}")
    print("")

File: cli-ts/src/lib/validate.ts
//@VALIDATE
import { readFileSync, existsSync } from 'node:fs';
import { join } from 'node:path';
import { parse } from 'yaml';

export interface ValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
  config?: {
    name: string;
    runtime: string;
    module: string;
    entrypoint: string;
    scaling?: { min_workers: number; max_workers: number };
  };
}

const VALID_RUNTIMES = ['python3', 'nodejs20'];

export function validateAgentYaml(agentPath: string): ValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  // Find agent.yaml
  let yamlPath = agentPath;
  if (!agentPath.endsWith('.yaml') && !agentPath.endsWith('.yml')) {
    yamlPath = join(agentPath, 'agent.yaml');
    if (!existsSync(yamlPath)) {
      yamlPath = join(agentPath, 'agent.yml');
    }
  }

  if (!existsSync(yamlPath)) {
    return {
      valid: false,
      errors: [`agent.yaml not found at ${agentPath}`],
      warnings: [],
    };
  }

  // Parse YAML
  let config: Record<string, unknown>;
  try {
    const content = readFileSync(yamlPath, 'utf-8');
    config = parse(content) as Record<string, unknown>;
  } catch (err) {
    return {
      valid: false,
      errors: [`Failed to parse YAML: ${err instanceof Error ? err.message : err}`],
      warnings: [],
    };
  }

  // Required fields
  if (!config.name || typeof config.name !== 'string') {
    errors.push('Missing required field: name');
  }

  if (!config.runtime || typeof config.runtime !== 'string') {
    errors.push('Missing required field: runtime');
  } else if (!VALID_RUNTIMES.includes(config.runtime)) {
    errors.push(`Invalid runtime: ${config.runtime}. Must be one of: ${VALID_RUNTIMES.join(', ')}`);
  }

  if (!config.module || typeof config.module !== 'string') {
    errors.push('Missing required field: module');
  }

  if (!config.entrypoint || typeof config.entrypoint !== 'string') {
    errors.push('Missing required field: entrypoint');
  }

  // Optional fields validation
  if (config.scaling) {
    const scaling = config.scaling as Record<string, unknown>;
    if (typeof scaling.min_workers !== 'number' || scaling.min_workers < 0) {
      warnings.push('scaling.min_workers should be a non-negative number');
    }
    if (typeof scaling.max_workers !== 'number' || scaling.max_workers < 1) {
      warnings.push('scaling.max_workers should be a positive number');
    }
    if (typeof scaling.min_workers === 'number' && typeof scaling.max_workers === 'number') {
      if (scaling.min_workers > scaling.max_workers) {
        errors.push('scaling.min_workers cannot be greater than scaling.max_workers');
      }
    }
  }

  if (config.memory_mb && (typeof config.memory_mb !== 'number' || config.memory_mb < 1)) {
    warnings.push('memory_mb should be a positive number');
  }

  if (config.timeout_seconds && (typeof config.timeout_seconds !== 'number' || config.timeout_seconds < 1)) {
    warnings.push('timeout_seconds should be a positive number');
  }

  return {
    valid: errors.length === 0,
    errors,
    warnings,
    config: errors.length === 0 ? {
      name: config.name as string,
      runtime: config.runtime as string,
      module: config.module as string,
      entrypoint: config.entrypoint as string,
      scaling: config.scaling as { min_workers: number; max_workers: number } | undefined,
    } : undefined,
  };
}
